src/ssl/support.cc

```diff
@@ -390,9 +390,17 @@ ssl_verify_cb(int ok, X509_STORE_CTX * ctx)
 }
 
 void
-Ssl::SetupVerifyCallback(Security::ContextPointer &ctx)
+Ssl::SetupPeerVerification(Security::ContextPointer &ctx, bool verify, bool fail_if_no_peer_cert)
 {
-    SSL_CTX_set_verify(ctx.get(), SSL_VERIFY_PEER | SSL_VERIFY_FAIL_IF_NO_PEER_CERT, ssl_verify_cb);
+  int mode;
+
+  mode = SSL_VERIFY_NONE;
+  if(verify)
```

Comment:
My reading of OpenSSL [docs](https://www.openssl.org/docs/man1.1.1/man3/SSL_CTX_set_verify.html) suggests that SSL_VERIFY_FAIL_IF_NO_PEER_CERT should have SSL_VERIFY_PEER set as well (emphasis mine):

> SSL_VERIFY_PEER [...] the server sends a client certificate request to the client. The certificate returned (if any) is checked. If the verification process fails, the TLS/SSL handshake is immediately terminated with an alert message containing the reason for the verification failure. The behaviour can be controlled by the **additional** SSL_VERIFY_FAIL_IF_NO_PEER_CERT, SSL_VERIFY_CLIENT_ONCE and SSL_VERIFY_POST_HANDSHAKE flags.

If I am right, then fixing this properly would require changing this function boolean parameters (and/or their handling) to make impossible (or reject) an invalid `(false, true)` combinations. I am not suggesting a specific fix because I hope that those boolean parameters will be gone as discussed in another change request.